### ISSUE TITLE
PPO best-response does not learn on bucket-brigade (root cause of #199/#215 non-convergence): value gradient swamps policy on shared trunk + dead grad-clip + BR undertraining

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/thrust#239]
## Summary

The PPO **best-response (BR) oracle does not learn on the bucket-brigade cells**, and this is the shared root cause of both the NFSP non-convergence (#199) and the PSRO non-convergence (#215). Surfaced empirically in the #134 full-scale re-run (48×8192, all 3 cells, 2026-06-22/23): with the #199/#215 knobs active (`AP_COVERAGE=2.0`, `BR_REWARD_SCALE=0.01`, `ALPHA_RANK_NORMALIZE_SPAN=1`), the BR stays at uniform-random and neither trainer converges. The micro-scale gains from #199/#215 did **not** generalize to the real budget.

## Empirical evidence (live #134 runs)

NFSP per-iteration BR diagnostics (cell beta05, iters 1–4):

| iter | avg_ap_loss | br pol | br val | br ent |
|------|-------------|--------|--------|--------|
| 1 | 3.6857 | −0.033 | 910.9 | 1.2295 |
| 2 | 3.6851 | −0.003 | 880.5 | 1.2292 |
| 3 | 3.6852 | +0.011 | 989.5 | 1.2290 |
| 4 | 3.6852 | −0.010 | 925.5 | 1.2287 |

- `avg_ap_loss` pinned at `ln(40)=3.689` (uniform floor) — reproduces #199 at scale.
- BR policy loss ≈ 0 and oscillating (no movement); BR value loss **flat at ~900** (critic never fits); BR entropy **pinned at exactly the 1.2297 per-dim uniform-max floor** (dead policy gradient).
- PSRO exploitability decreased for 3 iters (8356→7489→6660 — the #215 micro pattern) then **diverged** (→12840, oscillating 8k–14k) as the population filled with weak BRs — reproduces #215 at scale.

## Root-cause mechanism

The BR is a **shared-trunk actor-critic**: `fc1`/`fc2` are shared by the policy heads and the value head (`src/policy/multi_discrete_mlp.rs:30-34`, `forward` `:162-167`), trained with **one optimizer and one combined backward** `policy_loss + 0.5·value_loss − 0.01·entropy` (`src/multi_agent/joint.rs:783-821`).

1. **Value gradient swamps the policy on the shared trunk (primary).** Advantages are normalized (`joint.rs:698-702,904-913`) so the policy surrogate is ~0 with a tiny clipped gradient — *expected*, not the bug. The bug is that `0.5·value_loss` on raw-scale returns (~900) dominates the gradient into the shared `fc1`/`fc2`, dragging the features the policy heads read; policy logits never develop structure → entropy stuck at the `1.2297` floor.
2. **Gradient clipping is dead config.** `config.max_grad_norm = 0.5` exists (`joint.rs:308`) but is **never applied** — the joint step calls `optimizers[i].inner_mut().step(lr, policy, grads)` (`joint.rs:834`), bypassing `step_module`/`clip_grad_norm` (`src/train/optimizer.rs:116-123`), and the example optimizer factory `BurnOptimizer::new(AdamConfig::new().init(), 3e-4)` never calls `.clip_grad_norm(...)`. So nothing tames the value gradient.
3. **Severe BR undertraining.** The PPO update uses **one 256-sample minibatch per epoch** (not all minibatches — `joint.rs:710-717`, documented as a simplification at `joint.rs:48-55`), with `br_train_steps_per_iteration=1`. At `rollout_steps=8192`, ~97% of collected data is discarded each iteration; only `48 × 4 = 192` minibatch updates over a whole run — far too few to fit a hard critic regression + a 40-way joint-action policy against a moving opponent.
4. **Returns still large after 0.01 scaling.** Per-agent penalties reach −100 (`envs/bucket-brigade/bucket-brigade-core/src/scenarios.rs:1036-1062`), episodes ≥12 nights (`core.rs:437-446`); `0.01` shrinks the critic target but does not fix the actor/critic gradient imbalance.

## Config actually used (for reference)

actor+critic single Adam LR **3e-4** (shared); `vf_coef=0.5`, `ent_coef=0.01`, `gamma=0.99`, `gae_lambda=0.95`, `clip=0.2`, `clip_vf=0.2`, `n_epochs=4`, `minibatch_size=256`, `br_train_steps_per_iteration=1`, `normalize_advantages=true`, `max_grad_norm=0.5` (dead). (`train_nfsp.rs:220-246`, `train_psro.rs:292-324`, defaults `joint.rs:295-312`.)

## Ranked fixes (test in this order)

1. **Apply the already-configured gradient clipping** (wire `max_grad_norm` into the joint step / optimizer factory) **and lower `vf_coef` 0.5 → ~0.05** (already plumbed, `joint.rs:304`; expose as an env override on the examples).
2. **Iterate all minibatches per epoch** (`joint.rs:710-717` → loop over `generate_minibatch_indices_with_rng` chunks; helper exists at `src/.../loss.rs:214`) and/or expose `BR_TRAIN_STEPS` (default 1 → try 8).
3. **`BR_REWARD_SCALE=0.001`** (already an env knob) — apply together with #1.
4. (If still flat) **separate critic optimizer / higher value-head LR** (e.g. critic 1e-3, actor 3e-4).

## Acceptance criteria

- [ ] BR value loss falls well below 900 and trends down; BR entropy falls below 1.0 (policy moves off the uniform floor) within ~6 NFSP iters on one cell.
- [ ] NFSP `avg_ap_loss` demonstrably drops below `ln(40)=3.689`.
- [ ] (Gated on #134 cluster re-run) PSRO exploitability decreases over iterations on ≥1 cell, OR a documented finding that it still doesn't with a strong BR.
- [ ] A regression-safe default: changes either default to the improved behavior with passing tests, or are env/config-gated so existing tests/runs are unaffected.

## Fast A/B harness

No standalone single-cell BR trainer exists. Use a short NFSP run and watch the existing `br[val=… ent=…]` log (`train_nfsp.rs:292-314`):
```
TOTAL_ITERATIONS=6 ROLLOUT_STEPS=2048 CELL=beta05 BR_REWARD_SCALE=0.01 AP_COVERAGE=2.0 \
  cargo run --release --example train_nfsp --features "training,env-bucket-brigade"
```
(A purpose-built single-BR harness — freeze N−1 opponents, train one agent, log val-loss/entropy/explained-var — would be the right tool to build for tight A/B; it does not exist today.)

## Relevance

Root cause of #199 (NFSP no-learn) and #215 (PSRO non-convergence); the gating blocker for a positive #134 result; and #230 (α-rank perf) stays moot until a strong BR makes PSRO converge. Full analysis from the 2026-06-23 investigation during the #134 re-run.

**Comment by rjwalters** [comment]
## Update (2026-06-23): H1 ruled out — it's the critic, not gradient magnitude

PR #240 landed the infrastructure (wired the dead grad-clip, all-minibatch iteration, `VF_COEF`/`BR_TRAIN_STEPS` knobs) and ran the 6-iter NFSP A/B. **Decisive negative result that redirects the fix:**

| arm | br val (iter1→6) | br ent (iter6) | avg_ap_loss Δfloor |
|---|---|---|---|
| baseline (scale 0.01) | 888→868 | 1.2288 | −0.011 |
| fix: clip+all-mb+vf=0.05 | 878→867 | 1.2279 | −0.010 |
| fix + scale 0.001 | **8.1→8.1** | 1.2286 | −0.008 |

`BR_REWARD_SCALE=0.001` collapses BR value loss ~900→~8, so the value gradient **cannot** be swamping the shared trunk anymore — **yet entropy stays pinned at the 1.228 uniform-max floor and `avg_ap_loss` stays at ln(40).** So **H1 (gradient magnitude) is NOT the binding constraint.**

The new clue: at scale 0.001 the value loss is small but **flat** (8.08→7.60→8.11) — the critic is not fitting even tiny well-scaled targets. With normalized advantages, a non-fitting critic ⇒ advantages are noise ⇒ ~zero policy gradient ⇒ entropy stuck. **The binding constraint is the critic not fitting.**

### Refined next steps (the real fix is still open)
1. **Separate critic optimizer / higher value-head LR** (fix #4) — critic 1e-3 vs actor 3e-4, more critic steps.
2. **Is the value target even predictable?** Check for **partial observability** — can a per-agent local observation predict team/own return in this 4-player game? If not, the critic *can't* fit and PPO BR is structurally hard here (would need a centralized critic / different obs).
3. **Verify return/GAE correctness** for the joint multi-agent rollout (a bug here would make targets unfittable).
4. Network capacity (hidden=64) and explained-variance logging to make critic fit observable.

A fast single-BR harness (freeze N−1 opponents, log critic explained-variance per minibatch) would make this A/B much tighter — worth building as part of the next attempt.

**Comment by rjwalters** [comment]
### Follow-on issues filed (2026-06-23) — pick up here later

Broke the refined investigation into resumable threads (this issue stays the umbrella):

- **#243** — audit the joint rollout return/GAE computation for correctness (cheapest hypothesis to rule in/out — a target bug would explain the flat critic loss).
- **#241** — tooling: add critic **explained-variance** logging + a standalone single-BR A/B harness (the enabler that makes the rest fast).
- **#242** — the pivotal diagnostic: is the BR critic **structurally unfittable (partial observability)** → per-agent vs **centralized critic** decision.

Suggested order: #243 (rule out a bug) → #241 (tooling) → #242 (the decision). The #240 infra (wired grad-clip, all-minibatch, VF_COEF/BR_TRAIN_STEPS) is already merged and available.
